web/build.sh

```diff
@@ -0,0 +1,480 @@
+  ":web             Builds the website-oriented configuration of Keyman Engine for Web" \
+  ":ui              Builds the desktop UI modules used by the ${DOC_WEB_PRODUCT}" \
+  ":samples         Builds sample & test pages found under /samples and /testing, but not the ${DOC_WEB_PRODUCT}" \
+  ":tools           Builds related development + unit-test resources" \
+  "--skip-minify    Skips any minification steps in the build" \
+  "--all            Sets action to run on KMW's submodules as well if appropriate ($DOC_TEST_SYMBOL)"
+
+# Possible TODO?
+# "upload-symbols   Uploads build product to Sentry for error report symbolification.  Only defined for $DOC_BUILD_EMBED_WEB" \
+
+builder_describe_outputs \
+  configure         ../node_modules \
+  configure:embed   ../node_modules \
+  configure:web     ../node_modules \
+  configure:ui      ../node_modules \
+  configure:samples ../node_modules \
+  configure:tools   ../node_modules \
+  build:embed       $EMBED_OUTPUT/keyman.js \
+  build:web         $WEB_OUTPUT/keymanweb.js \
+  build:ui          $WEB_OUTPUT/kmwuibutton.js \
+  build:samples     $PREDICTIVE_TEXT_OUTPUT
+# Deliberately excluding build:tools b/c its script provides the definitions.
+
+builder_parse "$@"
+
+#### Build utility methods + definitions ####
+
+# Build products for each main target.
+WEB_TARGET=( "keymanweb.js" )
+UI_TARGET=( "kmwuibutton.js" "kmwuifloat.js" "kmwuitoggle.js" "kmwuitoolbar.js" )
+EMBED_TARGET=( "keyman.js" )
+
+: ${CLOSURECOMPILERPATH:=../node_modules/google-closure-compiler-java}
+: ${JAVA:=java}
+
+minifier="$CLOSURECOMPILERPATH/compiler.jar"
+
+# We'd love to add the argument --source_map_include_content for distribution in the future,
+# but Closure doesn't include the TS sources properly at this time.
+#
+# `checkTypes` is blocked b/c TypeScript can perform our type checking... and it causes an error
+# with TypeScript's `extend` implementation (it doesn't recognize a constructor without manual edits).
+# We also get a global `this` warning from the same.
+#
+# `checkVars` is blocked b/c Closure will otherwise fail on TypeScript namespacing, as each original TS
+# source file will redeclare the namespace variable, despite being merged into a single file post-compilation.
+#
+# `jsDocMissingType` prevents errors on type documentation Closure thinks is missing.  TypeScript may not
+# have the same requirements, and we trust TypeScript over Closure.
+minifier_warnings="--jscomp_error=* --jscomp_off=lintChecks --jscomp_off=unusedLocalVariables --jscomp_off=globalThis --jscomp_off=checkTypes --jscomp_off=checkVars --jscomp_off=jsdocMissingType --jscomp_off=uselessCode --jscomp_off=missingRequire --jscomp_off=strictMissingRequire"
+
+# We use these to prevent Closure from auto-inserting its own polyfills.  Turns out, they can break in the
+# WebView used by Android API 19, which our app still supports.
```

Comment:
Is this still applicable to our current minimum Android API 21

Reply:
I'd have to investigate.

Note that if we're able to achieve full conversion to ES Modules, many related bundling tools can also minify, such as `esbuild` - this would allow us to outright drop Google Closure.